File: apps/sandbox-runtime/internal/tunnel/agent_channel.go
package tunnel

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"time"

	"github.com/coder/websocket"
	"github.com/mistlehq/mistle/apps/sandbox-runtime/internal/httpclient"
	"github.com/mistlehq/mistle/apps/sandbox-runtime/internal/sessionprotocol"
	"github.com/mistlehq/mistle/apps/sandbox-runtime/internal/startup"
)

const agentEndpointDialTimeout = 5 * time.Second

type resolvedAgentEndpoint struct {
	RuntimeKey     string
	ClientID       string
	EndpointKey    string
	ConnectionMode string
	TransportURL   string
}

type agentRelayResult string

const (
	agentRelayResultDisconnected agentRelayResult = "disconnected"
)

func handleAgentConnectRequest(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	connectRequest connectRequest,
	executionLeases *executionLeaseEngine,
	agentRuntimes []startup.AgentRuntime,
	runtimeClients []startup.RuntimeClient,
	relayResultCh chan<- activeTunnelStreamRelayResult,
) (*activeTunnelStreamRelay, error) {
	agentEndpoint, err := resolveAgentEndpoint(agentRuntimes, runtimeClients)
	if err != nil {
		if writeErr := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
			Type:     sessionprotocol.MessageTypeStreamOpenError,
			StreamID: connectRequest.StreamID,
			Code:     connectErrorCodeAgentEndpointUnavailable,
			Message:  err.Error(),
		}); writeErr != nil {
			return nil, fmt.Errorf("failed to write sandbox tunnel stream.open error: %w", writeErr)
		}
		return nil, nil
	}
	if agentEndpoint == nil {
		if writeErr := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
			Type:     sessionprotocol.MessageTypeStreamOpenError,
			StreamID: connectRequest.StreamID,
			Code:     connectErrorCodeAgentEndpointUnavailable,
			Message:  "agent endpoint is not declared in runtime plan",
		}); writeErr != nil {
			return nil, fmt.Errorf("failed to write sandbox tunnel stream.open error: %w", writeErr)
		}
		return nil, nil
	}

	if agentEndpoint.ConnectionMode != "dedicated" {
		if err := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
			Type:     sessionprotocol.MessageTypeStreamOpenError,
			StreamID: connectRequest.StreamID,
			Code:     connectErrorCodeUnsupportedConnectionMode,
			Message:  fmt.Sprintf("connection mode '%s' is not supported", agentEndpoint.ConnectionMode),
		}); err != nil {
			return nil, fmt.Errorf("failed to write sandbox tunnel stream.open error: %w", err)
		}
		return nil, nil
	}

	agentConn, err := dialAgentEndpoint(ctx, agentEndpoint.TransportURL)
	if err != nil {
		if writeErr := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
			Type:     sessionprotocol.MessageTypeStreamOpenError,
			StreamID: connectRequest.StreamID,
			Code:     connectErrorCodeAgentEndpointDialFailed,
			Message:  err.Error(),
		}); writeErr != nil {
			return nil, fmt.Errorf("failed to write sandbox tunnel stream.open error: %w", writeErr)
		}
		return nil, nil
	}

	if err := writeStreamOpenOK(ctx, tunnelConn, sessionprotocol.StreamOpenOK{
		Type:     sessionprotocol.MessageTypeStreamOpenOK,
		StreamID: connectRequest.StreamID,
	}); err != nil {
		agentConn.CloseNow()
		return nil, fmt.Errorf("failed to write sandbox tunnel stream.open acknowledgement: %w", err)
	}

	observer := newAgentExecutionLeaseObserver(agentExecutionLeaseObserverInput{
		Context:         ctx,
		AgentRuntime:    startup.AgentRuntime{RuntimeKey: agentEndpoint.RuntimeKey},
		TransportURL:    agentEndpoint.TransportURL,
		ExecutionLeases: executionLeases,
	})

	return startAgentRelay(
		ctx,
		tunnelConn,
		agentConn,
		connectRequest.StreamID,
		observer,
		relayResultCh,
	), nil
}

func resolveAgentEndpoint(
	agentRuntimes []startup.AgentRuntime,
	runtimeClients []startup.RuntimeClient,
) (*resolvedAgentEndpoint, error) {
	if len(agentRuntimes) == 0 {
		return nil, nil
	}
	if len(agentRuntimes) > 1 {
		return nil, fmt.Errorf("runtime plan must declare at most one agent runtime for agent channel (found %d)", len(agentRuntimes))
	}

	agentRuntime := agentRuntimes[0]
	for _, runtimeClient := range runtimeClients {
		if runtimeClient.ClientID != agentRuntime.ClientID {
			continue
		}

		for _, endpoint := range runtimeClient.Endpoints {
			if endpoint.EndpointKey != agentRuntime.EndpointKey {
				continue
			}
			if endpoint.Transport.Type != "ws" {
				return nil, fmt.Errorf(
					"agent runtime '%s' on client '%s' must reference a websocket endpoint",
					agentRuntime.RuntimeKey,
					agentRuntime.ClientID,
				)
			}

			return &resolvedAgentEndpoint{
				RuntimeKey:     agentRuntime.RuntimeKey,
				ClientID:       runtimeClient.ClientID,
				EndpointKey:    endpoint.EndpointKey,
				ConnectionMode: endpoint.ConnectionMode,
				TransportURL:   endpoint.Transport.URL,
			}, nil
		}

		return nil, fmt.Errorf(
			"agent runtime '%s' references missing endpoint '%s' on client '%s'",
			agentRuntime.RuntimeKey,
			agentRuntime.EndpointKey,
			agentRuntime.ClientID,
		)
	}

	return nil, fmt.Errorf(
		"agent runtime '%s' references missing runtime client '%s'",
		agentRuntime.RuntimeKey,
		agentRuntime.ClientID,
	)
}

func dialAgentEndpoint(ctx context.Context, transportURL string) (*websocket.Conn, error) {
	dialContext, cancel := context.WithTimeout(ctx, agentEndpointDialTimeout)
	defer cancel()

	conn, _, err := websocket.Dial(dialContext, transportURL, &websocket.DialOptions{
		HTTPClient: httpclient.NewDirectClient(http.DefaultClient),
	})
	if err != nil {
		return nil, fmt.Errorf("failed to dial websocket endpoint %s: %w", transportURL, err)
	}

	return conn, nil
}

func relayTunnelFrames(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	agentConn *websocket.Conn,
	streamID int,
	observer agentExecutionLeaseObserver,
	incomingMessages <-chan tunnelMessage,
) (agentRelayResult, error) {
	relayContext, cancelAgentRelay := context.WithCancel(ctx)
	defer cancelAgentRelay()
	sendWindow := newStreamSendWindow()

	agentRelayErrCh := make(chan error, 1)
	go relayAgentFramesDirection(
		relayContext,
		agentRelayErrCh,
		agentConn,
		sendWindow,
		tunnelConn,
		uint32(streamID),
		observer,
	)

	for {
		select {
		case <-ctx.Done():
			return "", ctx.Err()
		case agentRelayErr := <-agentRelayErrCh:
			if agentRelayErr == nil || isExpectedAgentDisconnect(agentRelayErr) {
				return agentRelayResultDisconnected, nil
			}
			return "", agentRelayErr
		case message := <-incomingMessages:
			if message.MessageType == websocket.MessageText {
				controlMessageType, parseErr := parseControlMessageType(message.Payload)
				if parseErr == nil && controlMessageType == sessionprotocol.MessageTypeStreamWindow {
					streamWindow, windowErr := parseStreamWindow(message.Payload)
					if windowErr != nil {
						return "", windowErr
					}
					if streamWindow.StreamID != streamID {
						return "", fmt.Errorf(
							"stream.window streamId %d does not match active agent stream %d",
							streamWindow.StreamID,
							streamID,
						)
					}
					if err := sendWindow.add(streamWindow.Bytes); err != nil {
						if writeErr := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
							Type:     sessionprotocol.MessageTypeStreamReset,
							StreamID: streamID,
							Code:     streamResetCodeInvalidStreamWindow,
							Message:  err.Error(),
						}); writeErr != nil {
							return "", fmt.Errorf("failed to write stream.reset for excessive agent stream.window: %w", writeErr)
						}
						return agentRelayResultDisconnected, nil
					}
					continue
				}
				if parseErr == nil && controlMessageType == sessionprotocol.MessageTypeStreamClose {
					streamClose, closeErr := parseStreamClose(message.Payload)
					if closeErr != nil {
						return "", closeErr
					}
					if streamClose.StreamID != streamID {
						return "", fmt.Errorf(
							"stream.close streamId %d does not match active agent stream %d",
							streamClose.StreamID,
							streamID,
						)
					}
					return agentRelayResultDisconnected, nil
				}

				if err := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
					Type:     sessionprotocol.MessageTypeStreamReset,
					StreamID: streamID,
					Code:     streamResetCodeInvalidStreamData,
					Message:  "agent stream only accepts binary data frames after stream.open",
				}); err != nil {
					return "", fmt.Errorf("failed to write stream.reset for invalid agent text payload: %w", err)
				}
				return agentRelayResultDisconnected, nil
			}

			if message.MessageType != websocket.MessageBinary {
				return "", fmt.Errorf(
					"sandbox tunnel websocket produced unsupported message type %s",
					message.MessageType.String(),
				)
			}

			dataFrame, err := sessionprotocol.DecodeDataFrame(message.Payload)
			if err != nil {
				if writeErr := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
					Type:     sessionprotocol.MessageTypeStreamReset,
					StreamID: streamID,
					Code:     streamResetCodeInvalidStreamData,
					Message:  err.Error(),
				}); writeErr != nil {
					return "", fmt.Errorf("failed to write stream.reset for invalid agent data frame: %w", writeErr)
				}
				return agentRelayResultDisconnected, nil
			}
			if dataFrame.StreamID != uint32(streamID) {
				if err := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
					Type:     sessionprotocol.MessageTypeStreamReset,
					StreamID: streamID,
					Code:     streamResetCodeInvalidStreamData,
					Message:  fmt.Sprintf("stream data frame streamId %d does not match active agent stream %d", dataFrame.StreamID, streamID),
				}); err != nil {
					return "", fmt.Errorf("failed to write stream.reset for mismatched agent data frame: %w", err)
				}
				return agentRelayResultDisconnected, nil
			}

			var agentMessageType websocket.MessageType
			switch dataFrame.PayloadKind {
			case sessionprotocol.PayloadKindWebSocketText:
				agentMessageType = websocket.MessageText
			case sessionprotocol.PayloadKindWebSocketBinary:
				agentMessageType = websocket.MessageBinary
			default:
				if err := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
					Type:     sessionprotocol.MessageTypeStreamReset,
					StreamID: streamID,
					Code:     streamResetCodeInvalidStreamData,
					Message:  fmt.Sprintf("agent stream payloadKind %d is not supported", dataFrame.PayloadKind),
				}); err != nil {
					return "", fmt.Errorf("failed to write stream.reset for unsupported agent payload kind: %w", err)
				}
				return agentRelayResultDisconnected, nil
			}

			if err := agentConn.Write(ctx, agentMessageType, dataFrame.Payload); err != nil {
				if isExpectedAgentDisconnect(err) {
					return agentRelayResultDisconnected, nil
				}
				return "", fmt.Errorf("agent websocket write failed: %w", err)
			}
			if agentMessageType == websocket.MessageText && observer != nil {
				observer.ObserveClientMessage(dataFrame.Payload)
			}
			if err := writeStreamWindow(relayContext, tunnelConn, sessionprotocol.StreamWindow{
				Type:     sessionprotocol.MessageTypeStreamWindow,
				StreamID: streamID,
				Bytes:    len(dataFrame.Payload),
			}); err != nil {
				return "", fmt.Errorf("failed to write stream.window for consumed agent data: %w", err)
			}
		}
	}
}

func startAgentRelay(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	agentConn *websocket.Conn,
	streamID int,
	observer agentExecutionLeaseObserver,
	relayResultCh chan<- activeTunnelStreamRelayResult,
) *activeTunnelStreamRelay {
	relay := &activeTunnelStreamRelay{
		PrimaryStreamID: streamID,
		ChannelKind:     sessionprotocol.ChannelKindAgent,
		MessageCh:       make(chan tunnelMessage),
	}

	go func() {
		defer agentConn.CloseNow()

		result := activeTunnelStreamRelayResult{
			Relay: relay,
		}
		relayResult, err := relayTunnelFrames(ctx, tunnelConn, agentConn, streamID, observer, relay.MessageCh)
		if err != nil {
			result.Err = fmt.Errorf("sandbox tunnel websocket relay failed: %w", err)
		} else if relayResult == agentRelayResultDisconnected && observer != nil {
			observer.HandleStreamDisconnected()
		}
		relayResultCh <- result
	}()

	return relay
}

func relayAgentFramesDirection(
	ctx context.Context,
	relayErrCh chan<- error,
	source *websocket.Conn,
	sendWindow *streamSendWindow,
	target *websocket.Conn,
	streamID uint32,
	observer agentExecutionLeaseObserver,
) {
	for {
		messageType, payload, err := source.Read(ctx)
		if err != nil {
			relayErrCh <- fmt.Errorf("agent websocket read failed: %w", err)
			return
		}

		if messageType != websocket.MessageText && messageType != websocket.MessageBinary {
			relayErrCh <- fmt.Errorf("agent websocket produced unsupported message type %s", messageType.String())
			return
		}

		if !sendWindow.tryConsume(len(payload)) {
			if writeErr := writeStreamReset(ctx, target, sessionprotocol.StreamReset{
				Type:     sessionprotocol.MessageTypeStreamReset,
				StreamID: int(streamID),
				Code:     streamResetCodeStreamWindowExhausted,
				Message:  "agent stream send window is exhausted",
			}); writeErr != nil {
				relayErrCh <- fmt.Errorf("failed to write stream.reset for exhausted agent send window: %w", writeErr)
				return
			}
			relayErrCh <- nil
			return
		}

		var payloadKind byte
		switch messageType {
		case websocket.MessageText:
			payloadKind = sessionprotocol.PayloadKindWebSocketText
			if observer != nil {
				observer.ObserveAgentMessage(payload)
			}
		case websocket.MessageBinary:
			payloadKind = sessionprotocol.PayloadKindWebSocketBinary
		default:
			relayErrCh <- fmt.Errorf("agent websocket produced unsupported message type %s", messageType.String())
			return
		}

		if err := writeBinaryDataFrame(ctx, target, streamID, payloadKind, payload); err != nil {
			relayErrCh <- fmt.Errorf("sandbox tunnel websocket write failed: %w", err)
			return
		}
	}
}

func isExpectedAgentDisconnect(err error) bool {
	if err == nil {
		return false
	}

	if errors.Is(err, context.Canceled) {
		return true
	}

	switch websocket.CloseStatus(err) {
	case websocket.StatusNormalClosure, websocket.StatusGoingAway:
		return true
	default:
		return false
	}
}

File: apps/sandbox-runtime-node/src/tunnel/agent-channel.ts
import type {
  AgentExecutionObserverSession,
  CompiledAgentRuntime,
  CompiledRuntimeClient,
} from "@mistle/integrations-core";
import { resolveAgentExecutionObserver } from "@mistle/integrations-definitions/agent";
import {
  decodeDataFrame,
  PayloadKindWebSocketBinary,
  PayloadKindWebSocketText,
} from "@mistle/sandbox-session-protocol";
import type WebSocket from "ws";

import type { ActiveTunnelStreamRelay, ActiveTunnelStreamRelayResult } from "./active-relay.js";
import { trackObservedExecutions } from "./agent-execution-monitor.js";
import { AsyncQueue } from "./async-queue.js";
import type { TunnelSocketMessage } from "./connect-request.js";
import { parseControlMessageType, parseStreamCloseMessage } from "./connect-request.js";
import { ExecutionLeaseEngine } from "./execution-lease-engine.js";
import {
  CONNECT_ERROR_CODE_AGENT_ENDPOINT_DIAL_FAILED,
  CONNECT_ERROR_CODE_AGENT_ENDPOINT_UNAVAILABLE,
  CONNECT_ERROR_CODE_UNSUPPORTED_CONNECTION_MODE,
  STREAM_RESET_CODE_INVALID_STREAM_DATA,
  STREAM_RESET_CODE_INVALID_STREAM_WINDOW,
  STREAM_RESET_CODE_STREAM_WINDOW_EXHAUSTED,
  writeBinaryDataFrame,
  writeStreamOpenError,
  writeStreamOpenOk,
  writeStreamReset,
  writeStreamWindow,
} from "./messages.js";
import { StreamSendWindow } from "./stream-window.js";
import {
  connectWebSocket,
  closeWebSocket,
  createWebSocketMessageQueue,
  isExpectedWebSocketClose,
} from "./websocket.js";

export type ResolvedAgentEndpoint = {
  runtimeKey: string;
  clientId: string;
  endpointKey: string;
  adapterKey: string;
  connectionMode: "dedicated" | "shared";
  transportUrl: string;
};

export function resolveAgentEndpoint(
  agentRuntimes: ReadonlyArray<CompiledAgentRuntime>,
  runtimeClients: ReadonlyArray<CompiledRuntimeClient>,
): ResolvedAgentEndpoint | undefined {
  if (agentRuntimes.length === 0) {
    return undefined;
  }
  if (agentRuntimes.length > 1) {
    throw new Error(
      `runtime plan must declare at most one agent runtime for agent channel (found ${String(agentRuntimes.length)})`,
    );
  }

  const agentRuntime = agentRuntimes[0];
  if (agentRuntime === undefined) {
    throw new Error("agent runtime is required");
  }

  const runtimeClient = runtimeClients.find((client) => client.clientId === agentRuntime.clientId);
  if (runtimeClient === undefined) {
    throw new Error(
      `agent runtime '${agentRuntime.runtimeKey}' references missing runtime client '${agentRuntime.clientId}'`,
    );
  }

  const endpoint = runtimeClient.endpoints.find(
    (candidateEndpoint) => candidateEndpoint.endpointKey === agentRuntime.endpointKey,
  );
  if (endpoint === undefined) {
    throw new Error(
      `agent runtime '${agentRuntime.runtimeKey}' references missing endpoint '${agentRuntime.endpointKey}' on client '${agentRuntime.clientId}'`,
    );
  }
  if (endpoint.transport.type !== "ws") {
    throw new Error(
      `agent runtime '${agentRuntime.runtimeKey}' on client '${agentRuntime.clientId}' must reference a websocket endpoint`,
    );
  }

  return {
    runtimeKey: agentRuntime.runtimeKey,
    clientId: runtimeClient.clientId,
    endpointKey: endpoint.endpointKey,
    adapterKey: agentRuntime.adapterKey,
    connectionMode: endpoint.connectionMode,
    transportUrl: endpoint.transport.url,
  };
}

async function relayAgentFramesDirection(input: {
  signal: AbortSignal;
  agentSocket: WebSocket;
  observerSession: AgentExecutionObserverSession;
  sendWindow: StreamSendWindow;
  tunnelSocket: WebSocket;
  streamId: number;
}): Promise<void> {
  const agentMessages = createWebSocketMessageQueue(input.agentSocket);

  while (!input.signal.aborted) {
    let message: TunnelSocketMessage;
    try {
      message = await agentMessages.next(input.signal);
    } catch (error) {
      if (input.signal.aborted || isExpectedWebSocketClose(error)) {
        return;
      }

      throw new Error(
        `agent websocket read failed: ${error instanceof Error ? error.message : String(error)}`,
      );
    }

    const payloadBytes =
      message.kind === "text" ? new TextEncoder().encode(message.payload) : message.payload;
    input.observerSession.onInboundMessage(
      message.kind === "text" ? message.payload : message.payload,
    );
    if (!input.sendWindow.tryConsume(payloadBytes.length)) {
      await writeStreamReset(input.tunnelSocket, {
        type: "stream.reset",
        streamId: input.streamId,
        code: STREAM_RESET_CODE_STREAM_WINDOW_EXHAUSTED,
        message: "agent stream send window is exhausted",
      });
      return;
    }

    await writeBinaryDataFrame(input.tunnelSocket, {
      streamId: input.streamId,
      payloadKind: message.kind === "text" ? PayloadKindWebSocketText : PayloadKindWebSocketBinary,
      payload: payloadBytes,
    });
  }
}

async function relayTunnelFrames(input: {
  signal: AbortSignal;
  tunnelSocket: WebSocket;
  agentSocket: WebSocket;
  observerSession: AgentExecutionObserverSession;
  streamId: number;
  messages: AsyncQueue<TunnelSocketMessage>;
}): Promise<void> {
  const sendWindow = new StreamSendWindow();
  const outboundRelay = relayAgentFramesDirection({
    signal: input.signal,
    agentSocket: input.agentSocket,
    observerSession: input.observerSession,
    sendWindow,
    tunnelSocket: input.tunnelSocket,
    streamId: input.streamId,
  });

  while (!input.signal.aborted) {
    const nextTunnelMessage = input.messages.next(input.signal).then((message) => ({
      source: "tunnel" as const,
      message,
    }));
    const nextAgentResult = outboundRelay.then(() => ({
      source: "agent" as const,
    }));
    const nextEvent = await Promise.race([nextTunnelMessage, nextAgentResult]);

    if (nextEvent.source === "agent") {
      return;
    }

    const message = nextEvent.message;
    if (message.kind === "text") {
      const controlMessageType = parseControlMessageType(message.payload);
      if (controlMessageType === "stream.window") {
        const parsedWindow = JSON.parse(message.payload);
        const streamId =
          typeof parsedWindow.streamId === "number" && Number.isInteger(parsedWindow.streamId)
            ? parsedWindow.streamId
            : 0;
        const bytes =
          typeof parsedWindow.bytes === "number" && Number.isInteger(parsedWindow.bytes)
            ? parsedWindow.bytes
            : 0;
        if (streamId !== input.streamId) {
          throw new Error(
            `stream.window streamId ${String(streamId)} does not match active agent stream ${String(input.streamId)}`,
          );
        }
        try {
          sendWindow.add(bytes);
        } catch (error) {
          await writeStreamReset(input.tunnelSocket, {
            type: "stream.reset",
            streamId: input.streamId,
            code: STREAM_RESET_CODE_INVALID_STREAM_WINDOW,
            message: error instanceof Error ? error.message : String(error),
          });
          return;
        }
        continue;
      }
      if (controlMessageType === "stream.close") {
        const closeMessage = parseStreamCloseMessage(message.payload);
        if (closeMessage.streamId !== input.streamId) {
          throw new Error(
            `stream.close streamId ${String(closeMessage.streamId)} does not match active agent stream ${String(input.streamId)}`,
          );
        }
        return;
      }

      await writeStreamReset(input.tunnelSocket, {
        type: "stream.reset",
        streamId: input.streamId,
        code: STREAM_RESET_CODE_INVALID_STREAM_DATA,
        message: "agent stream only accepts binary data frames after stream.open",
      });
      return;
    }

    let dataFrame;
    try {
      dataFrame = decodeDataFrame(message.payload);
    } catch (error) {
      await writeStreamReset(input.tunnelSocket, {
        type: "stream.reset",
        streamId: input.streamId,
        code: STREAM_RESET_CODE_INVALID_STREAM_DATA,
        message: error instanceof Error ? error.message : String(error),
      });
      return;
    }
    if (dataFrame.streamId !== input.streamId) {
      await writeStreamReset(input.tunnelSocket, {
        type: "stream.reset",
        streamId: input.streamId,
        code: STREAM_RESET_CODE_INVALID_STREAM_DATA,
        message: `stream data frame streamId ${String(dataFrame.streamId)} does not match active agent stream ${String(input.streamId)}`,
      });
      return;
    }

    let outboundMessage: TunnelSocketMessage;
    switch (dataFrame.payloadKind) {
      case PayloadKindWebSocketText:
        outboundMessage = {
          kind: "text",
          payload: new TextDecoder().decode(dataFrame.payload),
        };
        break;
      case PayloadKindWebSocketBinary:
        outboundMessage = {
          kind: "binary",
          payload: dataFrame.payload,
        };
        break;
      default:
        await writeStreamReset(input.tunnelSocket, {
          type: "stream.reset",
          streamId: input.streamId,
          code: STREAM_RESET_CODE_INVALID_STREAM_DATA,
          message: `agent stream payloadKind ${String(dataFrame.payloadKind)} is not supported`,
        });
        return;
    }

    try {
      input.observerSession.onOutboundMessage(
        outboundMessage.kind === "text" ? outboundMessage.payload : outboundMessage.payload,
      );
      await connectAndSendAgentMessage(input.agentSocket, outboundMessage);
    } catch (error) {
      if (isExpectedWebSocketClose(error)) {
        return;
      }

      throw new Error(
        `agent websocket write failed: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
    await writeStreamWindow(input.tunnelSocket, {
      type: "stream.window",
      streamId: input.streamId,
      bytes: dataFrame.payload.length,
    });
  }
}

function connectAndSendAgentMessage(
  agentSocket: WebSocket,
  outboundMessage: TunnelSocketMessage,
): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    agentSocket.send(
      outboundMessage.payload,
      { binary: outboundMessage.kind === "binary" },
      (error) => {
        if (error == null) {
          resolve();
          return;
        }

        reject(error);
      },
    );
  });
}

export async function handleAgentConnectRequest(input: {
  signal: AbortSignal;
  tunnelSocket: WebSocket;
  streamId: number;
  agentRuntimes: ReadonlyArray<CompiledAgentRuntime>;
  runtimeClients: ReadonlyArray<CompiledRuntimeClient>;
  executionLeases: ExecutionLeaseEngine;
  executionLeasePollIntervalMs?: number;
  relayResultQueue: AsyncQueue<ActiveTunnelStreamRelayResult>;
}): Promise<ActiveTunnelStreamRelay | undefined> {
  let agentEndpoint: ResolvedAgentEndpoint | undefined;
  try {
    agentEndpoint = resolveAgentEndpoint(input.agentRuntimes, input.runtimeClients);
  } catch (error) {
    await writeStreamOpenError(input.tunnelSocket, {
      type: "stream.open.error",
      streamId: input.streamId,
      code: CONNECT_ERROR_CODE_AGENT_ENDPOINT_UNAVAILABLE,
      message: error instanceof Error ? error.message : String(error),
    });
    return undefined;
  }

  if (agentEndpoint === undefined) {
    await writeStreamOpenError(input.tunnelSocket, {
      type: "stream.open.error",
      streamId: input.streamId,
      code: CONNECT_ERROR_CODE_AGENT_ENDPOINT_UNAVAILABLE,
      message: "agent endpoint is not declared in runtime plan",
    });
    return undefined;
  }
  if (agentEndpoint.connectionMode !== "dedicated") {
    await writeStreamOpenError(input.tunnelSocket, {
      type: "stream.open.error",
      streamId: input.streamId,
      code: CONNECT_ERROR_CODE_UNSUPPORTED_CONNECTION_MODE,
      message: `connection mode '${agentEndpoint.connectionMode}' is not supported`,
    });
    return undefined;
  }

  let observerSession: AgentExecutionObserverSession;
  try {
    observerSession = resolveAgentExecutionObserver(agentEndpoint.adapterKey).createSession({
      transportUrl: agentEndpoint.transportUrl,
    });
  } catch (error) {
    await writeStreamOpenError(input.tunnelSocket, {
      type: "stream.open.error",
      streamId: input.streamId,
      code: CONNECT_ERROR_CODE_AGENT_ENDPOINT_UNAVAILABLE,
      message: error instanceof Error ? error.message : String(error),
    });
    return undefined;
  }

  let agentSocket: WebSocket;
  try {
    agentSocket = await connectWebSocket(agentEndpoint.transportUrl, input.signal);
  } catch (error) {
    await writeStreamOpenError(input.tunnelSocket, {
      type: "stream.open.error",
      streamId: input.streamId,
      code: CONNECT_ERROR_CODE_AGENT_ENDPOINT_DIAL_FAILED,
      message: error instanceof Error ? error.message : String(error),
    });
    return undefined;
  }

  await writeStreamOpenOk(input.tunnelSocket, {
    type: "stream.open.ok",
    streamId: input.streamId,
  });

  const relay: ActiveTunnelStreamRelay = {
    primaryStreamId: input.streamId,
    channelKind: "agent",
    messages: new AsyncQueue<TunnelSocketMessage>(),
  };

  void relayAgentSession({
    signal: input.signal,
    tunnelSocket: input.tunnelSocket,
    agentSocket,
    observerSession,
    executionLeases: input.executionLeases,
    ...(input.executionLeasePollIntervalMs === undefined
      ? {}
      : { executionLeasePollIntervalMs: input.executionLeasePollIntervalMs }),
    streamId: input.streamId,
    messages: relay.messages,
  })
    .then(() => {
      input.relayResultQueue.push({
        relay,
        updatesPtySession: false,
      });
    })
    .catch((error: unknown) => {
      input.relayResultQueue.push({
        relay,
        error: error instanceof Error ? error : new Error(String(error)),
        updatesPtySession: false,
      });
    });

  return relay;
}

async function relayAgentSession(input: {
  signal: AbortSignal;
  tunnelSocket: WebSocket;
  agentSocket: WebSocket;
  observerSession: AgentExecutionObserverSession;
  executionLeases: ExecutionLeaseEngine;
  executionLeasePollIntervalMs?: number;
  streamId: number;
  messages: AsyncQueue<TunnelSocketMessage>;
}): Promise<void> {
  try {
    await relayTunnelFrames(input);
  } finally {
    await closeWebSocket(input.agentSocket).catch(() => undefined);
    if (!input.signal.aborted) {
      trackObservedExecutions({
        signal: input.signal,
        executionLeases: input.executionLeases,
        observations: input.observerSession.drainObservations(),
        ...(input.executionLeasePollIntervalMs === undefined
          ? {}
          : { pollIntervalMs: input.executionLeasePollIntervalMs }),
      });
    }
  }
}
